### PR TITLE
Fix Metric.state_dict

### LINKED ISSUE #5603: The state of the child `Metric` is not saved to the parent module's `state_dict`.
## 🐛 Bug

The state of the child `Metric` is not saved to the parent module's `state_dict`.

### To Reproduce

```python
import pytorch_lightning as pl


class BoringModel(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.metric = pl.metrics.Accuracy()
        self.metric.persistent(mode=True)


model = BoringModel()

print(model.metric.state_dict())  # OrderedDict([('correct', tensor(0)), ('total', tensor(0))])
print(model.state_dict())         # OrderedDict()

```

### Expected behavior

The output should be
```console
OrderedDict([('correct', tensor(0)), ('total', tensor(0))])
OrderedDict([('metric.correct', tensor(0)), ('metric.total', tensor(0))])
```

### Environment

 - PyTorch Version (e.g., 1.0): 1.7.1+cpu
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.7.9
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information: pytorch-lightning: 1.1.5

### Additional context

1. The `args` and `kwargs` of `state_dict` are not passed to `super().state_dict`
https://github.com/PyTorchLightning/pytorch-lightning/blob/27e2b7f251b77271a1db12b7360ea6d784b65a85/pytorch_lightning/metrics/metric.py#L289

2. The `prefix` is not added to the `key`s of the states.
https://github.com/PyTorchLightning/pytorch-lightning/blob/27e2b7f251b77271a1db12b7360ea6d784b65a85/pytorch_lightning/metrics/metric.py#L293

Maybe we could change `Metric.state_dict` to
```python
    def state_dict(self, destination=None, prefix='', keep_vars=False):
        # Register metric states to be part of the state_dict
        destination = super().state_dict(
            destination=destination,
            prefix=prefix,
            keep_vars=keep_vars
        )
        for key in self._defaults.keys():
            if self._persistent[key]:
                current_val = getattr(self, key)
                if not keep_vars and torch.is_tensor(current_val):
                    current_val = current_val.detach()
                destination[prefix + key] = current_val
        return destination
```

### PR DESCRIPTION
**Pull request by manipopopo**
## What does this PR do?

Fixes #5603 :

```python
class BoringModel(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.metric = pl.metrics.Accuracy()
        self.metric.persistent(mode=True)
```
The state of `metric` is not added to the `state_dict` of `BoringModel`.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Fixed child `Metric` `state_dict` ([#5614](https://github.com/PyTorchLightning/pytorch-lightning/pull/5614))
```
Comment (SkafteNicki) [inline CHANGELOG.md:24]:
```suggestion
- Fixed `Metric`s `state_dict` not included when child modules  ([#5614](https://github.com/PyTorchLightning/pytorch-lightning/pull/5614))
```

File: pytorch_lightning/metrics/metric.py [hunk 2]
```diff
@@ -284,11 +284,17 @@ def persistent(self, mode: bool = False):
         for key in self._persistent.keys():
             self._persistent[key] = mode
 
-    def state_dict(self, *args, **kwargs):
+    def state_dict(self, destination=None, prefix='', keep_vars=False):
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:287]: I can see here https://github.com/pytorch/pytorch/blob/c7d348fea6ff4e20722256b38981c9c1f8727d95/torch/nn/modules/module.py#L1043-L1053 that `state_dict` is overloaded with two other signatures. Will this create problems?

File: pytorch_lightning/metrics/metric.py [hunk 3]
```diff
@@ -284,11 +284,17 @@ def persistent(self, mode: bool = False):
...
+        )
         # Register metric states to be part of the state_dict
-        state_dict = super().state_dict()
         for key in self._defaults.keys():
             if self._persistent[key]:
                 current_val = getattr(self, key)
-                state_dict.update({key: current_val})
-        return state_dict
+                if not keep_vars and torch.is_tensor(current_val):
+                    current_val = current_val.detach()
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:298]: should something similar be done when the state is a list?
```suggestion
                if not keep_vars:
                    if torch.is_tensor(current_val):
                        current_val = current_val.detach()
                    if isinstance(current_val, list) and all(torch.is_tensor(val) for val in current_val):
                        current_val = [val.detach() for val in current_val]
```

**Review by SkafteNicki — COMMENTED** [review]
some comments, but PR is looking good :]

File: CHANGELOG.md (hunk 1 above)
Comment (manipopopo) [reply CHANGELOG.md:24]: It's much clearer! Thank you.

File: pytorch_lightning/metrics/metric.py (hunk 3 above)
Comment (manipopopo) [reply pytorch_lightning/metrics/metric.py:298]: The _Note_ said
https://github.com/PyTorchLightning/pytorch-lightning/blob/6ab5417fd75f9e53ca5ceb28a7493da20a67adbb/pytorch_lightning/metrics/metric.py#L39-L41
Should we just check `isinstance(current_val, list)`:
```python
if isinstance(current_val, list):
    current_val = [val.detach() for val in current_val]
```
or do
```python
if isinstance(current_val, list):
    current_val = [val.detach() if torch.is_tensor(val) else val for val in current_val]
```
?

File: pytorch_lightning/metrics/metric.py (hunk 2 above)
Comment (manipopopo) [reply pytorch_lightning/metrics/metric.py:287]: https://github.com/pytorch/pytorch/commit/eace0533985641d9c2f36e43e3de694aca886bd9#diff-4be56271f7bfe650e3521c81fd363da58f109cd23ee80d243156d2d6ccda6263R746-R759
It seems that they are relevant to the deletion of the generation logic for the pyi files.

**Comment by Borda** [comment]
@teddykoker @SkafteNicki @justusschock mind re-review?

File: pytorch_lightning/metrics/metric.py (hunk 2 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:287]: okay, sound good to me :]

File: pytorch_lightning/metrics/metric.py (hunk 3 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/metric.py:298]: Lets do the second option. If an user implements a custom metric, they could store something else than tensors in the metric state (not that they should) so it is better to do an extra check at this point.

Comment (manipopopo) [reply pytorch_lightning/metrics/metric.py]: Thank you @SkafteNicki .
I've fixed it.

Following the naming of variables in https://github.com/PyTorchLightning/pytorch-lightning/blob/4edd07c1486f62c9f0bab4181840f2ea6eb7bfe0/pytorch_lightning/metrics/metric.py#L274
I change the variable name from `val` to `cur_v`.

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Review by justusschock — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
